packages/completer/src/widget.ts

```diff
@@ -725,30 +698,25 @@ export namespace Completer {
    * The default implementation of an `IRenderer`.
    */
   export class Renderer implements IRenderer {
-    /**
-     * Create an item node for a text completer menu.
-     */
     createItemNode(
-      item: IItem,
-      typeMap: TypeMap,
+      item: CompletionHandler.ICompletionItem,
       orderedTypes: string[]
     ): HTMLLIElement {
       let li = document.createElement('li');
       li.className = ITEM_CLASS;
       // Set the raw, un-marked up value as a data attribute.
-      li.setAttribute('data-value', item.raw);
+      li.setAttribute('data-value', item.insertText || item.label);
```

Comment:
This will work, but we may want to modify it slightly to make the future snippet support in extensions easier; to achieve that we may need more information from `ICompletionItem` available in the handler (rather than the `insertText` or `label` alone).

So, could we have an identifier (`ICompletionItem.id`) which would be replaced with the actual `ICompletionItem` object in wherever the `_selected` signal is emitted? Then the signal receiver ([`CompletionHandler.onCompletionSelected`](https://github.com/jupyterlab/jupyterlab/blob/2f049d57563b628b2e8d6202d716fb7bb7712687/packages/completer/src/handler.ts#L176)) would accept `item:  ICompletionItem` rather than `val: string`.

Form the extension point of view, the extension would only need to modify or re-implement `CompletionHandler` to get fancy new features such as snippets.

Reply:
On the other hand, if one wants to heavily modify the completer they will need to modify the renderer as well anyways...

Reply:
While this would certainly be a useful change, like pagination of completion items or asynchronous handling of documentation, this would probably be best served as a follow up. What do you think?

Reply:
I agree.